File: sdks/python/src/mcparmor/_process.py
"""ArmoredProcess — high-level interface for invoking MCP tools under armor."""

import io
import json
import logging
import select
import subprocess
from pathlib import Path
from types import TracebackType

_logger = logging.getLogger(__name__)

from mcparmor._popen import ArmorPopenError, armor_popen

# Sentinel used to detect reads that returned no data.
_EMPTY_LINE = ""


class ArmoredProcessError(OSError):
    """Raised when an armored process cannot be started or communication fails."""


class ArmoredProcess:
    """
    High-level interface for running an MCP tool under MCP Armor enforcement.

    Supports two usage patterns:

    **Single-call pattern** — the process is spawned per :meth:`invoke` call:

    .. code-block:: python

        proc = ArmoredProcess(command=["python", "tool.py"], armor="./armor.json")
        result = proc.invoke({"method": "run", "params": {}})

    **Persistent pattern** — the process is spawned once and reused:

    .. code-block:: python

        with ArmoredProcess(["npx", "-y", "@mcp/github"], armor="./armor.json") as proc:
            r1 = proc.invoke({"method": "list_repos", "params": {}})
            r2 = proc.invoke({"method": "get_issue", "params": {"number": 42}})
    """

    def __init__(
        self,
        command: list[str],
        *,
        armor: str | Path | None = None,
        profile: str | None = None,
        no_os_sandbox: bool = False,
        cwd: str | Path | None = None,
    ) -> None:
        """
        Initialise the ArmoredProcess configuration.

        The underlying subprocess is not started here; it is started either
        by entering the context manager or on the first :meth:`invoke` call
        when used outside a context manager.

        Args:
            command: The tool command and arguments to run under armor.
            armor: Path to the armor.json manifest.
            profile: Profile name override. Ignored if the manifest is locked.
            no_os_sandbox: Disable OS-level sandbox enforcement.
            cwd: Working directory for the spawned process.
        """
        self._command = command
        self._armor = armor
        self._profile = profile
        self._no_os_sandbox = no_os_sandbox
        self._cwd = cwd
        self._process: subprocess.Popen | None = None
        self._persistent = False

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    def __enter__(self) -> "ArmoredProcess":
        """Spawn the armored process and mark it as persistent."""
        self._persistent = True
        self._spawn()
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        """Terminate the armored process."""
        self.close()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def invoke(self, message: dict, *, timeout: float | None = None) -> dict:
        """
        Send a JSON-RPC message to the armored tool and return the parsed response.

        In single-call mode, a new subprocess is spawned for every call and
        terminated after the response is read. In persistent mode (context
        manager), the same process is reused across calls.

        Args:
            message: A JSON-serialisable dict representing the JSON-RPC request.
            timeout: Maximum seconds to wait for a response line. If exceeded,
                a :class:`TimeoutError` is raised.

        Returns:
            Parsed JSON-RPC response dictionary.

        Raises:
            ArmoredProcessError: If the process cannot be started or communication
                fails.
            TimeoutError: If no response is received within ``timeout`` seconds.
            json.JSONDecodeError: If the response line is not valid JSON.
        """
        if not self._persistent:
            return self._invoke_once(message, timeout=timeout)
        return self._invoke_persistent(message, timeout=timeout)

    def close(self) -> None:
        """
        Terminate the underlying subprocess if it is running.

        Safe to call multiple times; subsequent calls are no-ops.
        """
        if self._process is None:
            return
        try:
            self._process.terminate()
            self._process.wait(timeout=5)
        except OSError as exc:
            _logger.warning("Failed to terminate armored process cleanly: %s", exc)
        finally:
            self._process = None

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _spawn(self) -> None:
        """
        Start the armored subprocess with stdio pipes attached.

        Raises:
            ArmoredProcessError: If armor_popen raises ArmorPopenError.
        """
        try:
            self._process = armor_popen(
                self._command,
                armor=self._armor,
                profile=self._profile,
                no_os_sandbox=self._no_os_sandbox,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                cwd=self._cwd,
            )
        except ArmorPopenError as exc:
            raise ArmoredProcessError(str(exc)) from exc

    def _invoke_once(self, message: dict, *, timeout: float | None) -> dict:
        """Spawn, write, read, close — for single-call mode."""
        self._spawn()
        try:
            return self._send_and_receive(message, timeout=timeout)
        finally:
            self.close()

    def _invoke_persistent(self, message: dict, *, timeout: float | None) -> dict:
        """Write and read on the already-running process — for persistent mode."""
        if self._process is None:
            raise ArmoredProcessError("Process is not running; use as a context manager.")
        return self._send_and_receive(message, timeout=timeout)

    def _send_and_receive(self, message: dict, *, timeout: float | None) -> dict:
        """
        Write a JSON-RPC line to stdin and read one line from stdout.

        Args:
            message: JSON-serialisable request dictionary.
            timeout: Read timeout in seconds, or None for no timeout.

        Returns:
            Parsed response dictionary.

        Raises:
            ArmoredProcessError: If stdin/stdout are not available.
            TimeoutError: If the read exceeds ``timeout``.
        """
        proc = self._process
        if proc is None or proc.stdin is None or proc.stdout is None:
            raise ArmoredProcessError("Process stdio is unavailable.")

        line = json.dumps(message) + "\n"
        proc.stdin.write(line.encode())
        proc.stdin.flush()

        return _read_response(proc.stdout, timeout=timeout)


def _read_response(stdout: io.RawIOBase, *, timeout: float | None) -> dict:
    """
    Read one newline-delimited JSON line from stdout and parse it.

    Args:
        stdout: The process's stdout binary stream.
        timeout: Read deadline in seconds, or None for no limit.

    Returns:
        Parsed response dict.

    Raises:
        TimeoutError: If timeout is exceeded before a full line is received.
        ArmoredProcessError: If the line is empty (process closed stdout).
    """
    if timeout is not None:
        ready, _, _ = select.select([stdout], [], [], timeout)
        if not ready:
            raise TimeoutError("Timed out waiting for response from armored process.")

    raw = stdout.readline()
    text = raw.decode().strip() if isinstance(raw, bytes) else raw.strip()

    if text == _EMPTY_LINE:
        raise ArmoredProcessError("Armored process closed stdout without sending a response.")

    return json.loads(text)

File: sdks/node/src/process.ts
/**
 * ArmoredProcess — high-level wrapper for invoking MCP tools under armor enforcement.
 *
 * Manages the lifecycle of an mcparmor-brokered subprocess and provides a
 * simple JSON-RPC framing layer (line-delimited newline-terminated JSON) on
 * top of the process's stdin/stdout streams.
 */

import { type ChildProcess } from 'node:child_process';
import { armorSpawn, type ArmorSpawnOptions } from './spawn.ts';

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** Options accepted by {@link ArmoredProcess}. */
export interface ArmoredProcessOptions {
  /** The tool command and arguments (first element is the executable). */
  readonly command: readonly string[];
  /** Path to the armor.json capability manifest. */
  readonly armor?: string;
  /** Override the base profile declared in armor.json. */
  readonly profile?: string;
  /** Disable OS-level sandbox (skips Layer 2 enforcement). */
  readonly noOsSandbox?: boolean;
  /** Working directory for the spawned process. */
  readonly cwd?: string;
}

/** Options accepted by {@link ArmoredProcess.invoke}. */
export interface InvokeOptions {
  /** Maximum time in milliseconds to wait for a response. Default: 30 000. */
  readonly timeoutMs?: number;
}

// ---------------------------------------------------------------------------
// Error types
// ---------------------------------------------------------------------------

/** Thrown on process-level errors during spawn, I/O, or JSON parsing. */
export class ArmoredProcessError extends Error {
  constructor(message: string, options?: { cause?: unknown }) {
    super(message, options);
    this.name = 'ArmoredProcessError';
  }
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_TIMEOUT_MS = 30_000;

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * Build the ArmorSpawnOptions from ArmoredProcessOptions.
 *
 * Omits optional fields that are `undefined` to satisfy
 * `exactOptionalPropertyTypes`.
 *
 * @param options - The ArmoredProcessOptions to convert.
 * @returns Equivalent ArmorSpawnOptions for armorSpawn.
 */
function toSpawnOptions(options: ArmoredProcessOptions): ArmorSpawnOptions {
  return {
    stdio: 'pipe',
    ...(options.armor !== undefined ? { armor: options.armor } : {}),
    ...(options.profile !== undefined ? { profile: options.profile } : {}),
    ...(options.noOsSandbox !== undefined ? { noOsSandbox: options.noOsSandbox } : {}),
    ...(options.cwd !== undefined ? { cwd: options.cwd } : {}),
  };
}

/**
 * Write a JSON message followed by a newline to the process stdin.
 *
 * @param proc - The child process whose stdin to write to.
 * @param message - The message object to serialize.
 * @returns A promise that resolves when the write is flushed.
 * @throws {ArmoredProcessError} If stdin is not available or the write fails.
 */
function writeMessage(proc: ChildProcess, message: Record<string, unknown>): Promise<void> {
  return new Promise((resolve, reject) => {
    if (proc.stdin === null) {
      reject(new ArmoredProcessError('Process stdin is not available'));
      return;
    }
    const line = JSON.stringify(message) + '\n';
    const flushed = proc.stdin.write(line, (err) => {
      if (err !== undefined && err !== null) {
        reject(new ArmoredProcessError('Failed to write to process stdin', { cause: err }));
      }
    });
    if (flushed) {
      resolve();
    } else {
      proc.stdin.once('drain', resolve);
    }
  });
}

/**
 * Read exactly one newline-terminated line from stdout and parse it as JSON.
 *
 * @param proc - The child process to read from.
 * @param signal - AbortSignal used to cancel the read on timeout.
 * @returns Parsed JSON object from the response line.
 * @throws {ArmoredProcessError} If stdout is unavailable, the process exits
 *   before responding, the response is not valid JSON, or the signal fires.
 */
function readResponse(
  proc: ChildProcess,
  signal: AbortSignal,
): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    if (proc.stdout === null) {
      reject(new ArmoredProcessError('Process stdout is not available'));
      return;
    }

    let buffer = '';

    const onData = (chunk: Buffer | string): void => {
      buffer += chunk.toString();
      const newlineIndex = buffer.indexOf('\n');
      if (newlineIndex === -1) {
        return;
      }

      const line = buffer.slice(0, newlineIndex);
      cleanup();
      parseLine(line, resolve, reject);
    };

    const onClose = (): void => {
      cleanup();
      reject(new ArmoredProcessError('Process exited before sending a response'));
    };

    const onAbort = (): void => {
      cleanup();
      reject(new ArmoredProcessError('invoke() timed out waiting for a response'));
    };

    const cleanup = (): void => {
      proc.stdout?.removeListener('data', onData);
      proc.stdout?.removeListener('close', onClose);
      signal.removeEventListener('abort', onAbort);
    };

    proc.stdout.on('data', onData);
    proc.stdout.once('close', onClose);
    signal.addEventListener('abort', onAbort, { once: true });
  });
}

/**
 * Parse a JSON line and call the appropriate resolver or rejecter.
 *
 * @param line - The raw JSON string to parse.
 * @param resolve - Promise resolve callback.
 * @param reject - Promise reject callback.
 */
function parseLine(
  line: string,
  resolve: (value: Record<string, unknown>) => void,
  reject: (reason: ArmoredProcessError) => void,
): void {
  let parsed: unknown;
  try {
    parsed = JSON.parse(line);
  } catch (err) {
    reject(new ArmoredProcessError(`Process returned invalid JSON: ${line}`, { cause: err }));
    return;
  }

  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    reject(new ArmoredProcessError(`Process returned a non-object JSON value: ${line}`));
    return;
  }

  resolve(parsed as Record<string, unknown>);
}

/**
 * Create an AbortController that fires after timeoutMs milliseconds.
 *
 * The timer is deliberately kept referenced so it fires even when other
 * event-loop activity is absent. Callers must call controller.abort() in
 * a finally block to clear the timer promptly when the operation completes.
 *
 * @param timeoutMs - Duration in milliseconds before the signal fires.
 * @returns The AbortController; call .abort() to cancel before the timeout.
 */
function createTimeoutController(timeoutMs: number): AbortController {
  const controller = new AbortController();
  setTimeout(() => controller.abort(), timeoutMs);
  return controller;
}

// ---------------------------------------------------------------------------
// ArmoredProcess class
// ---------------------------------------------------------------------------

/**
 * Manages the lifecycle of an MCP tool subprocess running under MCP Armor.
 *
 * Supports two usage patterns:
 *
 * **Lazy spawn** (one-shot): each {@link invoke} call spawns a fresh process,
 * sends one message, reads one response, and exits.
 *
 * **Persistent spawn** (factory): use {@link ArmoredProcess.spawn} to eagerly
 * start the process. Subsequent {@link invoke} calls reuse the running process.
 * Call {@link close} when done.
 *
 * @example
 * ```typescript
 * // One-shot
 * const proc = new ArmoredProcess({ command: ['node', 'tool.js'], armor: './armor.json' });
 * const result = await proc.invoke({ method: 'list_repos', params: {} });
 *
 * // Persistent
 * const proc = await ArmoredProcess.spawn({ command: ['node', 'tool.js'], armor: './armor.json' });
 * const r1 = await proc.invoke(params1);
 * const r2 = await proc.invoke(params2);
 * await proc.close();
 * ```
 */
export class ArmoredProcess {
  readonly #options: ArmoredProcessOptions;
  #process: ChildProcess | null = null;

  /**
   * Create a new ArmoredProcess.
   *
   * The underlying subprocess is not started until the first {@link invoke}
   * call unless {@link ArmoredProcess.spawn} was used.
   *
   * @param options - Configuration for the armored process.
   */
  constructor(options: ArmoredProcessOptions) {
    this.#options = options;
  }

  /**
   * Factory that eagerly spawns the subprocess before the first invoke.
   *
   * Use this when you want to amortize startup latency across multiple invocations.
   *
   * @param options - Configuration for the armored process.
   * @returns A promise that resolves to a running {@link ArmoredProcess}.
   * @throws {ArmoredProcessError} If the process fails to start.
   */
  static async spawn(options: ArmoredProcessOptions): Promise<ArmoredProcess> {
    const instance = new ArmoredProcess(options);
    instance.#startProcess();
    return instance;
  }

  /**
   * Start the underlying subprocess if it is not already running.
   *
   * @throws {ArmoredProcessError} If the spawn fails.
   */
  #startProcess(): void {
    if (this.#process !== null) {
      return;
    }
    try {
      this.#process = armorSpawn(this.#options.command, toSpawnOptions(this.#options));
    } catch (err) {
      throw new ArmoredProcessError('Failed to spawn armored process', { cause: err });
    }
  }

  /**
   * Send a JSON-RPC message to the tool and return the parsed response.
   *
   * Writes `JSON.stringify(message) + "\n"` to stdin and reads one
   * newline-terminated line from stdout. The subprocess is started lazily
   * on the first call if not already running.
   *
   * @param message - The JSON-RPC message object to send.
   * @param options - Optional per-call settings (e.g. timeout).
   * @returns The parsed JSON response object from the tool.
   * @throws {ArmoredProcessError} On spawn failure, I/O error, invalid JSON
   *   response, or timeout.
   */
  async invoke(
    message: Record<string, unknown>,
    options: InvokeOptions = {},
  ): Promise<Record<string, unknown>> {
    this.#startProcess();

    const proc = this.#process as ChildProcess;
    const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    const controller = createTimeoutController(timeoutMs);

    try {
      await writeMessage(proc, message);
      return await readResponse(proc, controller.signal);
    } finally {
      controller.abort();
    }
  }

  /**
   * Terminate the subprocess and wait for it to exit.
   *
   * Safe to call when the process is not running.
   *
   * @returns A promise that resolves when the process has exited.
   */
  close(): Promise<void> {
    const proc = this.#process;
    if (proc === null) {
      return Promise.resolve();
    }

    this.#process = null;

    return new Promise((resolve) => {
      proc.once('close', resolve);
      proc.kill();
    });
  }
}
